### PR TITLE
Add support for per-route domains

### LINKED ISSUE #26846: Add domains to RateLimitPerRoute
*Title*: Add domains to RateLimitPerRoute

*Description*:
We want to set the domain in RLS requests on a per-route/per-vhost level.

This allows us to enable global rate limiting on our multi-tenant Envoy cluster, whilst allowing tenants to set unique policies on their particular hosts and routes.

I decided to take a crack at authoring the capability myself, PR is linked below.
It's been a long time since I've coded in C++ though so it'll likely need massaging/more tests.

I'm happy to drive the necessary changes, so please share feedback as required.

### PR DESCRIPTION
**Pull request by jacobneiltaylor**
Commit Message: Add support for per-route domains
Additional Description: Eanbles per-route configuration 
Risk Level: Low
Testing: Added new unit test to support the new case. I changed an existing unit test to make it build (removed InSequence), but would appreciate feedback on how to fix this.
Docs Changes: Added inline doco for API changes. Please suggest if more is required.
Release Notes: Updated `RateLimitPerRoute` with new `domain` field.
Platform Specific Features: Nil
Fixes #26846

### CODE REVIEW

**Comment by mattklein123** [comment]
Seems fine at a high level. Can you add a release note please?

/wait

**Comment by jacobneiltaylor** [comment]
Changeling has been added as requested.

**Review by tyxia — CHANGES_REQUESTED** [review]
Thanks for contribution!

Mostly LGTM, left one comment on API

File: api/envoy/extensions/filters/http/ratelimit/v3/rate_limit.proto [hunk 1]
```diff
@@ -385,4 +385,7 @@ message RateLimitPerRoute {
   // :ref:`RouteAction.rate_limits<envoy_v3_api_field_config.route.v3.RouteAction.rate_limits>` fields instead.
   // [#not-implemented-hide:]
   repeated RateLimitConfig rate_limits = 3;
+
+  // Overrides the domain. If not set, uses the filter-level domain instead.
+  string domain = 4 [(validate.rules).string = {min_len: 1}];
```
[the remark this answers is not in the document]
Comment (jacobneiltaylor) [reply api/envoy/extensions/filters/http/ratelimit/v3/rate_limit.proto]: Good catch. I have updated pushed an update to my dev branch that should remediate this.

File: source/extensions/filters/http/ratelimit/ratelimit.cc [hunk 2]
```diff
@@ -293,6 +293,19 @@ VhRateLimitOptions Filter::getVirtualHostRateLimitOption(const Router::RouteCons
...
+  std::string domain = config_->domain();
+  const auto* specific_per_route_config =
+      Http::Utility::resolveMostSpecificPerFilterConfig<FilterConfigPerRoute>(callbacks_);
+  if (specific_per_route_config != nullptr) {
+    std::string per_route_domain = specific_per_route_config->domain();
+    if (!per_route_domain.empty()) {
+      domain = per_route_domain;
+    }
+  }
+  return domain;
```
Comment (tyxia) [inline source/extensions/filters/http/ratelimit/ratelimit.cc:306]: nit: We can make this function a bit more concise and avoid some local variables like below:
```
std::string Filter::getDomain() { 
   const auto* specific_per_route_config =
      Http::Utility::resolveMostSpecificPerFilterConfig<FilterConfigPerRoute>(callbacks_);
   if (specific_per_route_config != nullptr && !specific_per_route_config->domain().empty()) {
      return specific_per_route_config->domain();
   }
   return config_->domain();
}
```

File: test/extensions/filters/http/ratelimit/ratelimit_test.cc [hunk 3]
```diff
@@ -1489,6 +1494,47 @@ TEST_F(HttpRateLimitFilterTest, IgnoreVHRateLimitOptionWithOutRouteRateLimit) {
       0, filter_callbacks_.clusterInfo()->statsScope().counterFromStatName(ratelimit_ok_).value());
 }
 
+// Tests that the domain is properly overridden when set at the per-route level
+TEST_F(HttpRateLimitFilterTest, PerRouteDomainSet) {
+  SetUpTest(filter_config_);
+  envoy::extensions::filters::http::ratelimit::v3::RateLimitPerRoute settings;
+  settings.set_domain("bar");
```
Comment (tyxia) [inline test/extensions/filters/http/ratelimit/ratelimit_test.cc:1501]: nit: can we have a local_variable like `const std::string per_route_domain = "bar"` And then, set it here and use it in limit() function at line#1515.

This is to highlight the check of per route domain name override, which is the key of this test.

**Review by tyxia — APPROVED** [review]
LGTM, modulo two cosmetic nits.

**Comment by tyxia** [comment]
@mattklein123 Could you take a look?  Thanks!

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**